File: LeetCode/Python/Hard/588-Design-In-Memory-File-System/solution.py
class TrieNode:
    def __init__(self, name: str, is_directory: bool):
        self.children=dict()

        self.name=name

        self.contents=""

        self.is_directory=is_directory

class Trie:
    def __init__(self):
        self.root=TrieNode("/", True)

    def get_directory_contents(self, path: str) -> List[str]:
        node=self.root

        for item in self.tokenize_path(path, "/"):
            if item not in node.children.keys():
                return []
            
            node=node.children[item]

        if not node.is_directory:
            return [node.name]

        return sorted([element.name for element in node.children.values()])

    def insert(self, path: str) -> None:
        node=self.root

        for item in self.tokenize_path(path, "/"):
            if item not in node.children.keys():
                node.children[item]=TrieNode(item, True)

            node=node.children[item]

    def append_content(self, path: str, file_contents: str) -> None:
        node=self.root

        for item in self.tokenize_path(path, "/"):
            if item not in node.children.keys():
                node.children[item]=TrieNode(item, True)

            node=node.children[item]

        node.is_directory=False

        node.contents+=file_contents

    def get_file_contents(self, path: str) -> str:
        node=self.root

        for item in self.tokenize_path(path, "/"):
            if item not in node.children.keys():
                return ""

            node=node.children[item]

        return node.contents

    def tokenize_path(self, path: str, delimiter: str) -> List[str]:
        tokens=[token for token in path.split(delimiter) if token]

        return tokens

class FileSystem:
    def __init__(self):
        self.root=Trie()

    def ls(self, path: str) -> List[str]:
        return self.root.get_directory_contents(path)

    def mkdir(self, path: str) -> None:
        self.root.insert(path)

    def addContentToFile(self, filePath: str, content: str) -> None:
        self.root.append_content(filePath, content)

    def readContentFromFile(self, filePath: str) -> str:
        return self.root.get_file_contents(filePath)

File: LeetCode/Rust/Hard/588-Design-In-Memory-File-System/solution.rs
use std::collections::{HashMap};

struct TrieNode
{
    children: HashMap<String, TrieNode>,

    name: String,

    contents: String,

    is_directory: bool
}

impl TrieNode
{
    fn new(name: String, is_directory: bool) -> Self
    {
        Self
        {
            children: HashMap::new(),

            name: name,

            contents: String::from(""),

            is_directory: is_directory
        }
    }
}

struct Trie
{
    root: TrieNode
}

impl Trie
{
    fn new() -> Self
    {
        Self
        {
            root: TrieNode::new("/".to_string(), true)
        }
    }

    fn get_directory_contents(&self, path: String) -> Vec<String>
    {
        let mut node: &TrieNode=&self.root;

        let tokens: Vec<String>=Self::tokenize_path(&path, '/');

        for token in tokens.iter()
        {
            if !node.children.contains_key(token)
            {
                return Vec::new();
            }

            node=node.children.get(token).unwrap();
        }

        if !node.is_directory
        {
            return vec![node.name.clone()];
        }

        let mut directory_contents: Vec<String>=Vec::new();

        for child in node.children.values()
        {
            directory_contents.push((*child).name.clone());
        }

        directory_contents.sort();

        directory_contents
    }

    fn insert(&mut self, path: String) -> ()
    {
        let mut node: &mut TrieNode=&mut self.root;

        let tokens: Vec<String>=Self::tokenize_path(&path, '/');

        for token in tokens.iter()
        {
            if !node.children.contains_key(token)
            {
                node.children.insert((*token).clone(), TrieNode::new((*token).clone(), true));
            }

            node=node.children.get_mut(token).unwrap();
        }
    }
    
    fn append_content(&mut self, path: String, file_contents: String) -> ()
    {
        let mut node: &mut TrieNode=&mut self.root;

        let tokens: Vec<String>=Self::tokenize_path(&path, '/');

        for token in tokens.iter()
        {
            if !node.children.contains_key(token)
            {
                node.children.insert((*token).clone(), TrieNode::new((*token).clone(), true));
            }

            node=node.children.get_mut(token).unwrap();
        }

        node.is_directory=false;

        node.contents.push_str(file_contents.as_str());
    }

    fn get_file_contents(&self, path: String) -> String
    {
        let mut node: &TrieNode=&self.root;

        let tokens: Vec<String>=Self::tokenize_path(&path, '/');

        for token in tokens.iter()
        {
            if !node.children.contains_key(token)
            {
                return String::from("");
            }

            node=node.children.get(token).unwrap();
        }

        node.contents.clone()
    }

    fn tokenize_path(path: &String, delimiter: char) -> Vec<String>
    {
        let tokens: Vec<String>=path.split(delimiter).filter(|part| !part.is_empty()).map(|part| part.to_string()).collect::<Vec<String>>();

        tokens
    }
}


struct FileSystem
{
    root: Trie
}

/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl FileSystem
{

    fn new() -> Self
    {
        Self
        {
            root: Trie::new()
        }
    }
    
    fn ls(&self, path: String) -> Vec<String>
    {
        self.root.get_directory_contents(path)
    }
    
    fn mkdir(&mut self, path: String)
    {
        self.root.insert(path)
    }
    
    fn add_content_to_file(&mut self, file_path: String, content: String)
    {
        self.root.append_content(file_path, content)
    }
    
    fn read_content_from_file(&self, file_path: String) -> String
    {
        self.root.get_file_contents(file_path)
    }
}